### ISSUE TITLE
Support for multiple buckets?

### ISSUE THREAD

**Issue by ffxsam** [aws-amplify/amplify-cli#5836]
In the documentation examples:

```js
Amplify.configure(
    Auth: {
        identityPoolId: 'XX-XXXX-X:XXXXXXXX-XXXX-1234-abcd-1234567890ab', //REQUIRED - Amazon Cognito Identity Pool ID
        region: 'XX-XXXX-X', // REQUIRED - Amazon Cognito Region
        userPoolId: 'XX-XXXX-X_abcd1234', //OPTIONAL - Amazon Cognito User Pool ID
        userPoolWebClientId: 'XX-XXXX-X_abcd1234', //OPTIONAL - Amazon Cognito Web Client ID
    },
    Storage: {
        bucket: '', //REQUIRED -  Amazon S3 bucket
        region: 'XX-XXXX-X', //OPTIONAL -  Amazon service region
    });
```

What if the web app needs to interact with more than one bucket? It would be nice to have a system where we could specify several and interact with them via their names.

```js
    Storage: {
      bucketOne: {
        bucket: '', //REQUIRED -  Amazon S3 bucket
        region: 'XX-XXXX-X', //OPTIONAL -  Amazon service region
      },
      bucketTwo: { ... }
    });
```

**Comment by sammartinez** [comment]
cc @renebrandel that we need Amplify CLI to do the implementation first prior to us doing anything on Amplify JS

**Comment by r0zar** [comment]
Does https://github.com/aws-amplify/amplify-cli/issues/3977 solve for this use case? I imagine that it would.

**Comment by stleon** [comment]
```
amplify import storage
Scanning for plugins...
Plugin scan successful
? Please select from one of the below mentioned services: S3 bucket - Content (Images, audio, video, etc.)
Amazon S3 storage was already added to your project.
```

It would be very useful if there were any support for multiple storages.

**Comment by arealmaas** [comment]
Ping 🙋

**Comment by santhedan** [comment]
This is a must have feature similar to how you allow multiple APIs to be invoked. Workaround exists for web but not for iOS and Android.

**Comment by nathanagez** [comment]
Any status on this feature ? :)

**Comment by johnrusch** [comment]
yeah, unable to list objects from another bucket using this workaround currently. Would love support for this feature.

**Comment by majirosstefan** [comment]
It would be nice if Amplify CLI supported **importing multiple buckets (even from different regions)** and maybe it would also ask which bucket should be "**the default one**" (I mean which bucket should be used when calling Storage methods without "bucket" param)

I also think that it would be really nice if there were some kind of official **blog posts**, Twitch streams, Youtube videos (whatever) about how to implement it ourselves (e.g. via patching) or how to create pull requests for AWS Amplify CLI in general (some kind of walkthrough of Amplify CLI).

Looking at https://github.com/aws-amplify/amplify-cli/tree/master/packages/amplify-cli it's not straightforward to me (a person outside Amplify team / mobile dev) how actually CLI works, and I am not sure if I want to spend time on this.

**Comment by rmjwilbur** [comment]
Agreed. It makes sense that one doesn't want to put all their eggs in one bucket. I would use this feature. Will explore workarounds for now. Thanks

**Comment by dragosiordachioaia** [comment]
How is this still not solved 4 years after the issue was created? So many real-life projects need to have more than one S3 buckets...

**Comment by macsint** [comment]
+1 to this, I'm starting to think that Amplify is very very limited when it comes to some real world scenarios, and I truly don't understand why these issues are not getting surfaced sooner. We like to brag about how scalable it can be, and how it could be used by large companies, but the more I use it, the more I realize that other than the authentication mechanism (which is a pain in the ass to do yourself), it is no scalable, and more suited for startup companies.

**Comment by macsint** [comment]
It is also baffling how, just as @dragosiordachioaia mentioned above that this issue was created 4 years ago and it was not yet addressed. I go back to Amazon's leadership principles of "Customer Obsession" and "Bias for Action" which are clearly not taken seriously here. :)

**Comment by annjawn** [comment]
I think it's a relatively straightforward implementation. In React projects I create a reusable configuration class which I initialize everytime I need to use Storage with the bucket and storage level I want. Something like this....

```javascript
//StorageService.js
import { Storage } from 'aws-amplify';
import { SetS3Config } from 'src/services/aws_services';

export default class StorageService{
    constructor(props){       
        const bucket = (props && props.bucket) ? props.bucket: '<default_bucket>';
        this.prefix =  (props && props.prefix) ? props.prefix : '';
        const config = (process.env.NODE_ENV === 'production')? configProdData :configDevData;
        Storage.configure({
             bucket,
             level : props.prefix ? props.prefix: 'public',  //this can be overridden at list, put, get level
             region: config.Auth.region,
             identityPoolId: config.Auth.identityPoolId
        });
    }

   // Other class methods for storage list, put, get, remove etc
   // You don't have to define these storage action methods here since Storage.configure 
   // overrides the configuration globally, but I like to keep everything pertaining to Storage together
   const list = async(key) => { 
            ....
            await Storage.list(key, {
              customPrefix: { public: this.prefix }, //you can also override protected and private prefixes, keeping it '' for all three will mean your app will have access to bucket's root, and ALL prefixes not just public/, private/, protected/ NOTE: make sure your auth and unauth IAM policies are set properly.
               pageSize : 'ALL'});
   }
   const upload = async(...) => { ... }
   ....
}
```

In the above, `configProdData` and `configDevData` is basically [Amplify configure JSON](https://docs.amplify.aws/lib/client-configuration/configuring-amplify-categories/q/platform/js/) (typically `amplify.prod.json` and `amplify.dev.json` that you import in `StorageService.js`). This is also where I define all my actions pertaining to Amplify Storage (`list`, `get`, `put` etc.). 

Now whenever I want to use "Storage Service" in a component or a custom Hook, all I do is 

```javascript
import StorageService from 'src/services/storage_services';

//initializes storage with <default_bucket>
const storage = new StorageService(); 
const listKeys = await storage.list(...); //gets the list of objects from <default_bucket> at public level or potentially all prefixes not just public/

//initializes storage with <default_bucket>
const storage = new StorageService({ level: 'protected' }); 
const listKeys = await storage.list(...); //gets the list of objects from <default_bucket> at protected level


//initializes storage with <some_other_bucket>
const storage = new StorageService({ bucket: '<some_other_bucket>' });
const listKeys = await storage.list(...); //gets the list of objects from <some_other_bucket> at public level or potentially all prefixes not just public/
```
and so on...

Another important thing to note is that all the buckets you want your app to access should be included in the IAM policy Auth_Role and Unauth_Role, for the Cognito IdentityPool otherwise it won't work (as defined [here](https://docs.amplify.aws/lib/storage/getting-started/q/platform/js/#using-amazon-s3)). Also, if you plan your app to access the bucket at other prefixes or even at the bucket's root level as defined by `customPrefix` your IAM policy should be set appropriately, not just with  `public/`, `private/`, `protected/` as the documentation shows, but with all the prefixes your app expects to access, or no prefix at all if you want your app to be able to access ALL prefixes under a bucket (although) not recommended just for security purposes).

Overall, I like this level of control rather than Amplify having an opinionated way of handling multiple buckets, which I am sure many will have different opinions about a native implementation. This method keeps the implementation flexibility on me so I can manage the Storage state at any given point simply by passing the bucket name and level. Some may disagree, but it has worked for me.

**Comment by majirosstefan** [comment]
One year later after my first post, but I got it working for mobile projects - React Native  ("aws-amplify": "^4.3.11",   "aws-amplify-react-native": "^6.0.2").

It took me few minutes to figure out that I need to edit roles, but it's working (in this case I did not need to call Storage.configure method).

```
  const RESOURCE_STORAGE_CONFIG = {
      level: "public" as StorageAccessLevel,
      bucket: RESOURCE_BUCKET_NAME,
      region: "us-east-2",
      expires: 60 * RESOURCE_EXPIRES_MINUTES,
    };

const resourceUrl =  await Storage.get(resourceS3Url, RESOURCE_STORAGE_CONFIG);

```

**Comment by charlieforward9** [comment]
Just read through this start to finish. I see there are some good workarounds, but it would definitely be nice to have some more up-to-date documentation rather than [this](https://docs.amplify.aws/cli/storage/import/#import-an-existing-s3-bucket) daunting quote:
> Amplify projects are limited to exactly one S3 bucket.
